### PR TITLE
NAS-119864 / 23.10 / Implement Available Apps header with filtering/searching

### PR DESCRIPTION
**Pull request by bvasilenko**
Please compare with UI mockups referenced in the ticket and test the functionality

### CODE REVIEW

**Review by undsoft — CHANGES_REQUESTED** [review]
1. The header is too cramped horizontally. We should have max-width on search block and we should add more free space between large numbers.
2. When filters are open, block with numbers jumps. It should stay locked to the top.
3. I think it's awkward to have to unselect app categories to filter by what you want. We should start with empty filter meaning all categories and let user select instead, preferably with autocomplete.
4. Design calls for a help text below app categories filter.
5. We can add unit tests for available-apps-header already.

**Comment by undsoft** [comment]
@dszidi We probably need to add a filter by trains here as well.

**Review by RehanY147 — CHANGES_REQUESTED** [review]
Please note that there is a new call `app.available` that we want to use to filter apps info. We want to use this endpoint instead of `catalog.query`. This endpoint also supports filters like the ones in the designs directly.

**Comment by bvasilenko** [comment]
@RehanY147 Ok, the requested changes are done

File: src/app/pages/apps/components/available-apps/available-apps-header/available-apps-header.component.spec.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        mockCall('app.available', [{
+          categories: ['storage', 'crypto'],
+          last_update: '2023-03-01 13:26:19',
+          name: 'chia',
+        }, {
+          categories: ['media', 'torrent'],
+          last_update: '2023-02-28 16:37:54',
+          name: 'qbittorent',
+        }] as unknown as AvailableApp[]),
+        mockCall('chart.release.query', [{}, {}, {}] as unknown as ChartRelease[]),
```
Comment (undsoft) [inline src/app/pages/apps/components/available-apps/available-apps-header/available-apps-header.component.spec.ts:30]: Casting to unknown is unnecessary here and a line above.

File: src/app/pages/apps/components/available-apps/available-apps-header/available-apps-header.component.spec.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    spectator = createComponent();
+  });
+
+  it('checks the displayed numbers', () => {
+    const numbers = spectator.queryAll('.header-number h2');
+
+    expect(numbers[0]).toHaveText('2'); // available apps
+    expect(numbers[1]).toHaveText('3'); // installed apps
+    expect(numbers[2]).toHaveText('1'); // installed catalogs
+  });
```
Comment (undsoft) [inline src/app/pages/apps/components/available-apps/available-apps-header/available-apps-header.component.spec.ts:45]: Component does a bit more than just showing 3 numbers.

File: src/app/pages/apps/services/applications.service.ts [hunk 3]
```diff
@@ -26,6 +29,28 @@ export class ApplicationsService {
     return this.ws.call('catalog.query', [[], { extra: { cache: true, item_details: true } }]);
   }
 
+  getAvailableItem(name: string, catalog: string, train: string): Observable<AvailableApp> {
+    const firstOption: QueryFilter<AvailableApp>[] = [];
+    firstOption.push(['name', '=', name]);
+    firstOption.push(['catalog', '=', catalog]);
+    firstOption.push(['train', '=', train]);
+    return this.ws.call('app.available', [firstOption]).pipe(switchMap((app) => of(app[0])));
```
Comment (undsoft) [inline src/app/pages/apps/services/applications.service.ts:37]: Or you know, just `map` :)

**Review by undsoft — APPROVED** [review]
Generally looks okay.

ix-chips had a stroke on me.
Снимок экрана 2023-03-16 в 14 15 30

**Review by RehanY147 — APPROVED** [review]
Please address feedback by @undsoft

**Pull request merged**